File: src/main/java/io/myosotisdev/minestom/entity/PlayerEx.java
package io.myosotisdev.minestom.entity;

import io.myosotisdev.minestom.permission.IPermissionHandler;
import io.myosotisdev.minestom.permission.Permission;
import io.myosotisdev.minestom.permission.PermissionGroup;
import net.minestom.server.entity.Player;
import net.minestom.server.network.player.PlayerConnection;
import net.minestom.server.permission.PermissionVerifier;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

public class PlayerEx
        extends Player
        implements IPermissionHandler
{
    private Set<Permission> permissions;
    private PermissionGroup permGroup;

    public PlayerEx(@NotNull UUID uuid, @NotNull String username, @NotNull PlayerConnection playerConnection)
    {
        super(uuid, username, playerConnection);
        this.permissions = new HashSet<>();
    }

    @Override
    public boolean hasPermission(@NotNull net.minestom.server.permission.Permission permission)
    {
        return permission instanceof Permission && hasPermission((Permission) permission);
    }

    @Deprecated
    @Override
    public boolean hasPermission(@NotNull String permissionName, @Nullable PermissionVerifier permissionVerifier)
    {
        return this.hasPermission(permissionName);
    }

    @Override
    public boolean hasPermission(@NotNull String permissionName)
    {
        return this.hasPermission(Permission.ofString(permissionName));
    }

    @Override
    public void addPermission(Permission permission)
    {
        this.permissions.add(permission);
    }

    @Override
    public void removePermission(Permission permission)
    {
        this.permissions.remove(permission);
    }

    @Override
    public boolean hasPermission(Permission permission)
    {
        return (permGroup != null ?
                permGroup.hasPermission(permission) :
                false) ||
                permissions.stream()
                           .anyMatch(perm -> perm.matchWith(permission));
    }

    @Override
    public Collection<Permission> permissions()
    {
        return permissions;
    }
}

File: src/main/java/io/myosotisdev/minestom/entity/PlayerEx.kt
package io.myosotisdev.minestom.entity

import io.myosotisdev.minestom.permission.IPermissionHandler
import io.myosotisdev.minestom.permission.Permission
import io.myosotisdev.minestom.permission.Permission.Companion.ofString
import io.myosotisdev.minestom.permission.PermissionGroup
import net.minestom.server.entity.Player
import net.minestom.server.network.player.PlayerConnection
import java.util.*

open class PlayerEx(uuid: UUID, username: String, playerConnection: PlayerConnection) : Player(uuid, username, playerConnection),
                                                                                        IPermissionHandler
{
    private val permissions: MutableSet<Permission?>
    private val permGroup: PermissionGroup? = null

    init
    {
        this.permissions = HashSet()
    }

    override fun hasPermission(permission: net.minestom.server.permission.Permission): Boolean
    {
        return permission is Permission && hasPerm(permission)
    }

    override fun hasPermission(permissionName: String): Boolean
    {
        return this.hasPerm(ofString(permissionName))
    }

    override fun addPerm(permission: Permission?)
    {
        this.permissions.add(permission)
    }

    override fun removePerm(permission: Permission?)
    {
        this.permissions.remove(permission)
    }

    override fun hasPerm(permission: Permission?): Boolean
    {
        return permGroup?.hasPerm(permission) ?: false ||
               permissions.stream()
                       .anyMatch { perm: Permission? -> perm!!.matchWith(permission!!) }
    }

    override fun permissions(): Collection<Permission?>?
    {
        return permissions
    }
}